Feat/gh 73/crowdactions

edreinoso opened the pull request:
Included crowdaction endpoint functionality in the following structure:

- handler/crowdaction.go
- internals/crowdaction.go
- models/crowdaction.go | commitment.go
- repository/aws/dynamo.go
- repository/aws/crowdactionManager.go

edreinoso committed:
🚀 crowdaction hex refactoring init

edreinoso committed:
🚧  dynamodb manager repository, pt1: init

edreinoso committed:
⚡️ launching own sam config

edreinoso committed:
Merge branch 'development' into feat/gh-73/crowdactions

edreinoso committed:
Merge branch 'development' into feat/gh-73/crowdactions

edreinoso committed:
🚧 [crowdaction] deployed first function on to aws

edreinoso committed:
🚧 [crowdaction] removing extra copy of dynamo repository

edreinoso committed:
🚧 [crowdaction] init model and validation on request

edreinoso committed:
🚧 [crowdaction] changed the model function names

edreinoso committed:
🚧 [crowdaction] working on internal and handler

edreinoso committed:
🚧 [crowdaction] simple version of GET request by ID

edreinoso committed:
🚧 [crowdaction] testing GET request by ID

edreinoso committed:
🚧 [crowdaction] creating external function to GET crowdaction by ID

edreinoso committed:
🚧 [crowdaction] including query function in ddb repository

edreinoso committed:
🚧 [crowdaction] initial code base for GET crowdactions by status

edreinoso committed:
🚧 [crowdaction] cleaning up, commitment options

rubenhorn commented on samconfig.yaml:
```diff
@@ -0,0 +1,12 @@
+version = 0.1
```
This file should not be checked into git

edreinoso replied:
Should I just remove it? for future commits, or how do I get rid of it from the history?

rubenhorn replied:
The information is not too sensitive, but this file is unique to each deployment, so it should not be tracked in git.
So you can just create a new commit that removes it (and maybe add it to the `.gitignore`, if it is not yet properly ignored)

edreinoso replied:
Perfect, done!

rubenhorn commented on pkg/handler/aws/emailContact/main.go:
```diff
@@ -22,6 +24,8 @@ func handler(ctx context.Context, req events.APIGatewayV2HTTPRequest) (events.AP
 	}
 	// TODO implement POW verification using nonce (see https://github.com/CollActionteam/collaction_backend/issues/58)
 
+	fmt.Println("Hello World")
```
Please remove test code

rubenhorn commented on internal/crowdactions/crowdaction.go:
```diff
@@ -0,0 +1,81 @@
+	"github.com/aws/aws-sdk-go/service/dynamodb/expression"
+)
+
+const Separator = "### app version:"
```
Remnant from copy paste from other file? 😉  
Please remove

edreinoso replied:
Yes it was a copy past from another file 😄 . It's been removed

rubenhorn commented on pkg/repository/aws/dynamo.go:
```diff
@@ -71,6 +74,8 @@ func (s *Dynamo) GetPrimaryKey(pk string, sk string) PrimaryKey {
 }
 
 func (s *Dynamo) GetDBItem(tableName string, pk string, sk string) (map[string]*dynamodb.AttributeValue, error) {
+	// fmt.Println("Calling GetDBItem", tableName, pk, sk)
```
Please remove commented out code

rubenhorn commented on pkg/repository/aws/dynamo.go:
```diff
@@ -87,9 +92,62 @@ func (s *Dynamo) GetDBItem(tableName string, pk string, sk string) (map[string]*
 		return nil, nil
 	}
+
+	// fmt.Println("result from query: ", result)
```
Please remove commented out code

rubenhorn commented on pkg/repository/aws/dynamo.go:
```diff
@@ -87,9 +92,62 @@ func (s *Dynamo) GetDBItem(tableName string, pk string, sk string) (map[string]*
 }
 
+func (s *Dynamo) Query(tableName string, filterCond expression.ConditionBuilder, startFrom *utils.PrimaryKey) ([]models.CrowdactionData, error) {
+	// var crowdactions models.CrowdactionData
```
Please remove commented out code

rubenhorn commented on pkg/handler/aws/crowdaction/main.go:
```diff
@@ -0,0 +1,93 @@
+	fmt.Println("Hello Go, first message from CollAction AWS:  ", crowdactionID)
+	var request models.CrowdactionRequest
+
+	//dynamoRepository := awsRepository.NewDynamo() // should pass this dynamo variable
```
Please remove commented out code

rubenhorn commented on internal/crowdactions/crowdaction.go:
```diff
@@ -0,0 +1,81 @@
+	GET Crowdaction by Id
+**/
+func (e *crowdaction) GetCrowdactionById(ctx context.Context, crowdactionID string) (*models.CrowdactionData, error) {
+	// fmt.Println("GetCrowdaction calling internal:", crowdactionID)
```
Please remove commented out code

rubenhorn commented on internal/crowdactions/crowdaction.go:
```diff
@@ -0,0 +1,81 @@
+	switch status {
+	case "joinable":
+		// call joinable crowdaction list
+		// crowdactions, _, err = models.ListJoinableCrowdactions(tableName, nil) // TODO future
```
Please remove commented out code

rubenhorn commented on internal/crowdactions/crowdaction.go:
```diff
@@ -0,0 +1,81 @@
+	"github.com/CollActionteam/collaction_backend/internal/constants"
+	"github.com/CollActionteam/collaction_backend/internal/models"
+	"github.com/CollActionteam/collaction_backend/utils"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
```
We should avoid using aws dependencies outside of the aws repositories.

edreinoso replied:
This is a very good point, the thing is that these dependencies are used by the crowdaction's internal. If this is the case, should I move this piece of logic somewhere else?

For example: 
The expression package is being used for the filter condition

`
"github.com/aws/aws-sdk-go/service/dynamodb/expression"
`
`
filterCond := expression.Name(KeyDateJoinBefore).GreaterThan(expression.Value(utils.GetDateStringNow()))
`

rubenhorn replied:
Yeah.. that is indeed an issue.
I'm not sure if it's worth it to create stand-in objects to be used outside of the aws repositories for everything.
Maybe we should make an exception for DynamoDB, since we will (probably) never use a different database and thus tight coupling might be more permissible. (?)
@mrsoftware what is your view of this?

mrsoftware replied:
logic should not know anything about the AWS or any external dependencies, instead, we should create an adaptor in this case repository also work as an adaptor, so logic doesn't care about AWS or any external source.

so if we need an extra function/type/ to decuple AWS from handler or logic, it's okay to write/add it.

rubenhorn commented on internal/crowdactions/crowdaction.go:
```diff
@@ -0,0 +1,81 @@
+
+const Separator = "### app version:"
+
+type DynamoRepository interface {
```
The interface should only use non-aws types.  
(Probably should also be called something like `CrowdactionRepository` or `DatabaseRepository`, but I am not sure about this part of the architecture. Maybe @mrsoftware has an opinion about this.)

edreinoso replied:
Oh okay, maybe I was confused but I thought they would be the same functions as in the repository. 

For example if you look at the internal/contact/email.go, you can see that EmailRepository is declaring a Send func, which is then being used in the SendEmail function.

This Send func is the same as there is in the in the pkg/repository/aws/emailManager.go file.

rubenhorn replied:
Yes. `internal/contact/...` contains the services with the interfaces for repositories (ports) and `pkg/repository/...` contains the corresponding adapters.

mrsoftware replied:
yes, the logic does not care if we use dynamo/Redis/MySQL, the logic defines its dependences and we provide it.
so `CrowdactionRepository` is more proper than `DynamoRepository`

edreinoso replied:
Perfect then, I understand now. This was an issue on my side for misinterpreting the naming convention for the specific type. I have fixed it now.

rubenhorn reviewed (changes requested):
* Please remove empty file `pkg/handler/aws/crowdaction/README.md`
* Please add mock for repository
* Please add unit tests (using mock)

edreinoso committed:
🚧 [crowdaction] including the mock repository for dynamo and crowdaction

edreinoso committed:
🚧 [crowdaction] modifying the mock repository to include GetCrowdactionByStatus

edreinoso committed:
🚧 [crowdaction] removed unnecessary comments, per the feedback on pull request

mrsoftware commented on internal/crowdactions/crowdaction.go:
```diff
@@ -0,0 +1,81 @@
+const Separator = "### app version:"
+
+type DynamoRepository interface {
+	GetDBItem(tableName string, pk string, crowdactionId string) (map[string]*dynamodb.AttributeValue, error)
```
``the repository should not export adapted types,
you can alias the type or create a specific type and convert  `dynamodb.AttributeValue` to it.



note: creating an alias for an external type is like this:

```Go
type AttributeValue = dynamodb.AttributeValue
```

but creating an internal type and converting it is better and less error-prone.

mrsoftware commented on internal/crowdactions/crowdaction.go:
```diff
@@ -0,0 +1,81 @@
+		// call joinable crowdaction list
+		// crowdactions, _, err = models.ListJoinableCrowdactions(tableName, nil) // TODO future
+		filterCond := expression.Name(KeyDateJoinBefore).GreaterThan(expression.Value(utils.GetDateStringNow()))
+		items, err := e.dynamodb.Query(constants.TableName, filterCond, startFrom)
```
doing query over a database is repository responsibility, so you should create a separate repository for `Crowdactions` and move this logic to it, in this method we only call the `CrowdactionsRepository.GetByStatus(prams...)`

mrsoftware commented on pkg/handler/aws/crowdaction/main.go:
```diff
@@ -0,0 +1,93 @@
+
+	if err != nil {
+		return utils.CreateMessageHttpResponse(http.StatusInternalServerError, err.Error()), nil
+	} else {
```
useless else, if it is an error, it returns and otherwise continuing

mrsoftware commented on pkg/handler/aws/crowdaction/main.go:
```diff
@@ -0,0 +1,93 @@
+func handler(ctx context.Context, req events.APIGatewayV2HTTPRequest) (events.APIGatewayV2HTTPResponse, error) {
+	crowdactionID := req.PathParameters["crowdactionID"]
+
+	fmt.Println("Hello Go, first message from CollAction AWS:  ", crowdactionID)
```
remove debug comments.

edreinoso committed:
[pull request] changes done based on feedback

[pull request] ignoring samconfig file as per the pull request

[pull request] modified naming for the internal interface

[pull request] getting rid of useless else statement

[pull request] removing dynamodb dependencies in logic

[pull request] including mock repository and crowdaction test

[pull request] cleaning up some unnecessary lines

[pull request] separating repository concerns

[pull request] modifying test cases with new repository structure

[pull request] changing dynamo repository call from crowdaction main handler

[pull request] changing parameters to pass unit test case

[pull request] changing crowdactionID from test case

[pull request] changing method call for crowdaction test

[pull request] fixing parameter issue in crowdaction_test

[pull request] testing issue with the args.get(x) call in mock/dynamo repository

[pull request] including Mock in each of the function calls from the dynamo repository

[pull request] modifying test cases for crowdaction

edreinoso committed:
[pull request] removing samconfig file from history

edreinoso committed:
[pull request] ignoring samconfig file as per the pull request

edreinoso committed:
[pull request] modified naming for the internal interface

edreinoso committed:
[pull request] getting rid of useless else statement

edreinoso committed:
[pull request] removing dynamodb dependencies in logic

edreinoso committed:
[pull request] including mock repository and crowdaction test

edreinoso committed:
[pull request] cleaning up some unnecessary lines

rubenhorn commented on .gitignore:
```diff
@@ -37,4 +37,6 @@ _ignored/
\ No newline at end of file
+.idea
+
+samconfig.yaml
```
Ah.. now I know why it was checked in in the first place.  
Isn't this ususally a `.toml` file?

edreinoso replied:
What do you mean? I am not sure if I got your point

rubenhorn commented on pkg/repository/aws/dynamo.go:
```diff
@@ -73,9 +73,8 @@ func (s *Dynamo) GetPrimaryKey(pk string, sk string) PrimaryKey {
-func (s *Dynamo) GetDBItem(tableName string, pk string, sk string) (map[string]*dynamodb.AttributeValue, error) {
-	// fmt.Println("Calling GetDBItem", tableName, pk, sk)
-
+func (s *Dynamo) GetDBItem(tableName string, pk string, sk string) (*models.CrowdactionData, error) {
```
If we return `*models.CrowdactionData` I think the method name should reflect that, since we will need to return other objects from DynamoDB as well.

edreinoso replied:
You have an excellent point, this was done in one of the commits but then modified in the final version that was committed for review.

rubenhorn commented on pkg/repository/aws/dynamo.go:
```diff
@@ -93,12 +92,12 @@ func (s *Dynamo) GetDBItem(tableName string, pk string, sk string) (map[string]*
-func (s *Dynamo) Query(tableName string, filterCond expression.ConditionBuilder, startFrom *utils.PrimaryKey) ([]models.CrowdactionData, error) {
-	// var crowdactions models.CrowdactionData
+func (s *Dynamo) Query(tableName string, filter string, startFrom *utils.PrimaryKey) ([]models.CrowdactionData, error) {
+	filterCond := expression.Name(filter).GreaterThan(expression.Value(utils.GetDateStringNow()))
```
Can we be sure that it is always `> now` in the filter`?  
For "ended" this does not seem to be the case.  
Maybe these things should also be covered by unit tests?

rubenhorn commented on internal/crowdactions/crowdaction_test.go:
```diff
@@ -0,0 +1,31 @@
+	crowdactionID := "Helloworld2"
+
+	t.Run("dev stage", func(t *testing.T) {
+		dynamoRepository.On("Send", context.Background(), crowdactionID).Return(nil).Once()
```
What is this "Send" function? I don't see it anywhere inside the dynamoRepository or mock.

edreinoso replied:
This was modified in later commits.

rubenhorn reviewed (changes requested):
[Build is failing!](https://github.com/CollActionteam/collaction_backend/runs/5683660347?check_suite_focus=true#step:6:36)

edreinoso committed:
[pull request] separating repository concerns

edreinoso committed:
[pull request] modifying test cases with new repository structure

edreinoso committed:
[pull request] changing dynamo repository call from crowdaction main handler

edreinoso committed:
[pull request] changing parameters to pass unit test case

edreinoso committed:
[pull request] changing crowdactionID from test case

edreinoso committed:
[pull request] changing method call for crowdaction test

edreinoso committed:
[pull request] fixing parameter issue in crowdaction_test

edreinoso committed:
[pull request] testing issue with the args.get(x) call in mock/dynamo repository

edreinoso committed:
[pull request] including Mock in each of the function calls from the dynamo repository

edreinoso committed:
[pull request] modifying test cases for crowdaction

edreinoso committed:
Merge branch 'feat/gh-73/crowdactions' of github.com:CollActionteam/collaction_backend into feat/gh-73/crowdactions

rubenhorn commented on internal/crowdactions/crowdaction.go:
```diff
@@ -0,0 +1,48 @@
+}
+
+/**
+	GET Crowdaction by Id
```
Comment seems a bit redundant

edreinoso replied:
Will get rid of it

rubenhorn commented on internal/crowdactions/crowdaction.go:
```diff
@@ -0,0 +1,48 @@
+}
+
+/**
+	GET Crowdaction by Status
```
Comment seems a bit redundant

rubenhorn commented on pkg/repository/aws/crowdactionManager.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+/**
+	GET Crowdaction by Status
```
Comment seems a bit redundant

edreinoso replied:
Will get rid of it

rubenhorn commented on pkg/repository/aws/crowdactionManager.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+/**
+	GET Crowdaction by Id
```
Comment seems a bit redundant

edreinoso replied:
Will get rid of this

rubenhorn commented on internal/crowdactions/crowdaction_test.go:
```diff
@@ -0,0 +1,36 @@
+
+		crowdaction, err := service.GetCrowdactionById(ctx, crowdactionID)
+
+		fmt.Println("Hello world", crowdaction)
```
Can this be removed?

edreinoso replied:
We could, but then this would give me an error that the crowdaction variable is not being used. What do we do in this case?

vrazdalovschi replied:
> We could, but then this would give me an error that the crowdaction variable is not being used. What do we do in this case?

It depends on your test. If you want to test, that `GetCrowdactionById` logic works, then you should assert that your result has valid field values. If you don't care about crowdaction result, then write:

`_, err := service.GetCrowdactionById(ctx, crowdactionID)`

edreinoso replied:
Thanks, that was helpful!

mrsoftware commented on pkg/handler/aws/emailContact/main.go:
```diff
@@ -22,6 +24,8 @@ func handler(ctx context.Context, req events.APIGatewayV2HTTPRequest) (events.AP
 	}
 	// TODO implement POW verification using nonce (see https://github.com/CollActionteam/collaction_backend/issues/58)
 
+	fmt.Println("Hello World")
```
pls remove this line

edreinoso replied:
Will get rid of it

edreinoso committed:
[pull request] removing unnecessary hello world

edreinoso committed:
[pull request] removing some unnecessary comments

edreinoso committed:
Merge branch 'development' of github.com:CollActionteam/collaction_backend into feat/gh-73/crowdactions

vrazdalovschi commented:
Hi, I am not familiar with the whole project, but why do you use a lot of `fmt.Print`? 
Like:
`fmt.Println("GetCrowdactionsByStatus", status, startFrom)`

Is it part of debug logs or for local testing during development? If it is done for log purposes, then all these prints must be replaced with a logger.

vrazdalovschi commented:
@edreinoso could you please rebase your branch to reduce MERGE commits in your PR?

edreinoso committed:
final changes before merging

This commit starts by containing the separation between the post and the get methods that will be used for handling crowdactions

edreinoso committed:
Merge branch 'development' of github.com:CollActionteam/collaction_backend into feat/gh-73/crowdactions

# Conflicts:
#	pkg/handler/aws/emailContact/main.go

edreinoso committed:
removing a duplicate package

When merged, the net/http packe was imported twice by accident

rubenhorn commented on pkg/handler/aws/crowdaction/crowdaction.go:
```diff
@@ -0,0 +1,94 @@
+}
+
+func (c *CrowdactionHandler) createCrowdaction(ctx context.Context, req events.APIGatewayV2HTTPRequest) (events.APIGatewayV2HTTPResponse, error) {
+	// validate request
```
Please mark with `TODO`, so it will be easier to find with _Ctrl+Shift+F_

rubenhorn commented on pkg/handler/aws/crowdaction/crowdaction.go:
```diff
@@ -0,0 +1,94 @@
+}
+
+func (c *CrowdactionHandler) getCrowdactionByID(ctx context.Context, crowdactionID string) (events.APIGatewayV2HTTPResponse, error) {
+	// func getCrowdactionByID(ctx context.Context, crowdactionID string) (events.APIGatewayV2HTTPResponse, error) {
```
Please remove commented out code or add `TODO` + explanation if it may still be needed later

edreinoso replied:
Yes, I am going to clean this up a bit more and then merge

edreinoso committed:
[pull request] cleaning up some final comments

This should be the last commit before merging to development.

The implementation includes the initial basis for the /post endpoint to create crowdactions from the CMS.

The continuation of this task (/post) should be done ideally in the CMS task (up for a discussion, of course) but could expanded somewhere else for other purposes.

For example, the badge issue may implement a very rudimentary version of this /post crowdaction for testing purposes.

edreinoso committed:
changing the crowdaction test case
